File: Runtime/Plugins/Android/PerkoxUnityBridge.java
package com.perkox.unity;

import android.app.Activity;
import android.util.Log;
import com.unity3d.player.UnityPlayer;
import com.perkoxofferwall.sdk.PerkoxOfferwall;
import com.perkoxofferwall.sdk.Offerwall;
import org.json.JSONObject;
import java.util.HashMap;
import java.util.Map;
import kotlin.Unit;
import kotlin.jvm.functions.Function0;
import kotlin.jvm.functions.Function1;

/**
 * PerkoxUnityBridge
 * Android Java bridge for Perkox Unity SDK.
 */
public class PerkoxUnityBridge {
    private static final String TAG = "PerkoxUnity";
    private static final String RECEIVER_OBJECT = "PerkoxCallbackReceiver";

    private static String activeAppId = "";
    private static String activeSdkKey = "";
    private static String activePlayerId = "";
    private static boolean activeBeta = false;

    public static void initSDK(String appId, String sdkKey, String playerId, boolean beta) {
        activeAppId = appId != null ? appId.trim() : "";
        activeSdkKey = sdkKey != null ? sdkKey.trim() : "";
        activePlayerId = playerId != null ? playerId.trim() : "";
        activeBeta = beta;
        Log.d(TAG, "Initialized PerkoxUnityBridge (appId=" + activeAppId + ", beta=" + activeBeta + ")");
    }

    public static void setUserId(String playerId) {
        activePlayerId = playerId != null ? playerId.trim() : "";
        Log.d(TAG, "Updated Player ID to: " + activePlayerId);
    }

    public static void showOfferwall(final Activity activity, String appId, String sdkKey, String playerId, final boolean beta) {
        if (activity == null) {
            Log.e(TAG, "Cannot show offerwall: Activity is null");
            UnityPlayer.UnitySendMessage(RECEIVER_OBJECT, "OnOfferwallError", "Activity is null");
            return;
        }

        final String app = (appId != null && !appId.isEmpty()) ? appId.trim() : activeAppId;
        final String key = (sdkKey != null && !sdkKey.isEmpty()) ? sdkKey.trim() : activeSdkKey;
        final String user = (playerId != null && !playerId.isEmpty()) ? playerId.trim() : activePlayerId;

        if (app.isEmpty() || key.isEmpty()) {
            String err = "Cannot show offerwall: appId and sdkKey must not be empty";
            Log.e(TAG, err);
            UnityPlayer.UnitySendMessage(RECEIVER_OBJECT, "OnOfferwallError", err);
            return;
        }

        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                try {
                    Offerwall offerwall = PerkoxOfferwall.INSTANCE.create(app, key, user);

                    offerwall.setOnReward(new Function1<Map<String, ?>, Unit>() {
                        @Override
                        public Unit invoke(Map<String, ?> rewardMap) {
                            try {
                                JSONObject json = new JSONObject(rewardMap != null ? rewardMap : new HashMap<String, Object>());
                                UnityPlayer.UnitySendMessage(RECEIVER_OBJECT, "OnRewardReceivedInternal", json.toString());
                            } catch (Exception ex) {
                                Log.e(TAG, "Error serializing reward callback JSON: " + ex.getMessage());
                                UnityPlayer.UnitySendMessage(RECEIVER_OBJECT, "OnRewardReceivedInternal", "{}");
                            }
                            return Unit.INSTANCE;
                        }
                    });

                    offerwall.setOnClose(new Function0<Unit>() {
                        @Override
                        public Unit invoke() {
                            UnityPlayer.UnitySendMessage(RECEIVER_OBJECT, "OnOfferwallClosedInternal", "");
                            return Unit.INSTANCE;
                        }
                    });

                    offerwall.launch(activity, beta);
                    UnityPlayer.UnitySendMessage(RECEIVER_OBJECT, "OnOfferwallOpenedInternal", "");
                } catch (Throwable t) {
                    String msg = t.getMessage() != null ? t.getMessage() : "Unknown error launching offerwall";
                    Log.e(TAG, "Exception launching offerwall: " + msg, t);
                    UnityPlayer.UnitySendMessage(RECEIVER_OBJECT, "OnOfferwallError", msg);
                }
            }
        });
    }
}

File: Runtime/Plugins/iOS/PerkoxUnityBridge.swift
import Foundation
import UIKit

#if canImport(PerkoxOfferwall)
import PerkoxOfferwall
#endif

// Forward declaration of UnitySendMessage from Unity runtime
@_silgen_name("UnitySendMessage")
func UnitySendMessage(_ obj: UnsafePointer<CChar>?, _ method: UnsafePointer<CChar>?, _ msg: UnsafePointer<CChar>?)

@objc public class PerkoxUnityBridge: NSObject {
    private static var activeAppId: String = ""
    private static var activeSdkKey: String = ""
    private static var activePlayerId: String = ""
    private static var activeBeta: Bool = false

    private static let receiverObject = "PerkoxCallbackReceiver"

    private static func sendToUnity(method: String, message: String) {
        receiverObject.withCString { objPtr in
            method.withCString { methodPtr in
                message.withCString { msgPtr in
                    UnitySendMessage(objPtr, methodPtr, msgPtr)
                }
            }
        }
    }

    @objc public static func initSDK(_ appId: String, sdkKey: String, playerId: String, beta: Bool) {
        activeAppId = appId.trimmingCharacters(in: .whitespacesAndNewlines)
        activeSdkKey = sdkKey.trimmingCharacters(in: .whitespacesAndNewlines)
        activePlayerId = playerId.trimmingCharacters(in: .whitespacesAndNewlines)
        activeBeta = beta
    }

    @objc public static func setUserId(_ playerId: String) {
        activePlayerId = playerId.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    @objc public static func showOfferwall(_ appId: String, sdkKey: String, playerId: String, beta: Bool) {
        DispatchQueue.main.async {
            guard let topVC = getTopViewController() else {
                sendToUnity(method: "OnOfferwallError", message: "Unable to find top UIViewController to present offerwall")
                return
            }

            #if canImport(PerkoxOfferwall)
            let app = !appId.isEmpty ? appId.trimmingCharacters(in: .whitespacesAndNewlines) : activeAppId
            let key = !sdkKey.isEmpty ? sdkKey.trimmingCharacters(in: .whitespacesAndNewlines) : activeSdkKey
            let user = !playerId.isEmpty ? playerId.trimmingCharacters(in: .whitespacesAndNewlines) : activePlayerId

            if app.isEmpty || key.isEmpty {
                sendToUnity(method: "OnOfferwallError", message: "Cannot show offerwall: appId and sdkKey must not be empty")
                return
            }

            let offerwall = PerkoxOfferwall.create(appId: app, sdkKey: key, playerId: user)

            offerwall.onReward = { reward in
                do {
                    var sanitized: [String: Any] = [:]
                    for (k, v) in reward {
                        if let val = v {
                            sanitized[k] = val
                        }
                    }
                    let jsonData = try JSONSerialization.data(withJSONObject: sanitized, options: [])
                    if let jsonString = String(data: jsonData, encoding: .utf8) {
                        sendToUnity(method: "OnRewardReceivedInternal", message: jsonString)
                    }
                } catch {
                    sendToUnity(method: "OnRewardReceivedInternal", message: "{}")
                }
            }

            offerwall.onClose = {
                sendToUnity(method: "OnOfferwallClosedInternal", message: "")
            }

            offerwall.launch(viewController: topVC, beta: beta)
            sendToUnity(method: "OnOfferwallOpenedInternal", message: "")
            #else
            sendToUnity(method: "OnOfferwallError", message: "PerkoxOfferwall framework is not linked in Unity iOS build")
            #endif
        }
    }

    private static func getTopViewController(from rootVC: UIViewController? = nil) -> UIViewController? {
        let root = rootVC ?? {
            if #available(iOS 13.0, *) {
                return UIApplication.shared.connectedScenes
                    .compactMap { $0 as? UIWindowScene }
                    .flatMap { $0.windows }
                    .first(where: { $0.isKeyWindow })?.rootViewController
            } else {
                return UIApplication.shared.keyWindow?.rootViewController
            }
        }()

        if let nav = root as? UINavigationController {
            return getTopViewController(from: nav.visibleViewController)
        }
        if let tab = root as? UITabBarController {
            return getTopViewController(from: tab.selectedViewController)
        }
        if let presented = root?.presentedViewController {
            return getTopViewController(from: presented)
        }
        return root
    }
}
